55、（多选题）如下代码执行时会抛异常的有（BD）

A、result = {(1, 2), 1, 2}

B、result = {[1, 2], 1, 2}

C、result = ([1, 2], 1, 2)

D、result = {[1, 2] : 12, 1 : 1, 2 : 2}

可变类型有列表、集合、字典
{}表示集合，集合中的元素相当于字典中的键
所以，虽然集合是可变类型，但是集合的元素必须是不可变类型
()为元组，不可变
A:
    {}表示集合，集合只能放不可变类型
B:
    列表是可变类型，报错
C:
    元组可以放多种类型，可变不可变都可以，并且元组内的可变类型是可以修改的
D:
    字典属于可变类型，但是字典的键不能是可变类型


```python
class B():
    def __init__(self):
        print('B')


class A(B):
    def __init__(self):
        ____________ # 需填入代码
```

a = A()
以下哪些选项中的代码单独填入到横线处，可以使得以上代码段输出B（AD）

A、super().__init__()

B、super().__init__(self)

C、B.__init__()

D、B.__init__(self)

这里很显然，A继承了B，super().__init__()相当于执行B的__init__，输出B
直接调用B的函数显然也是可以的
BC属于语法错误

In [545]:
# 相当于int("12")+2
print(int("1" + "2") + 2)

14


In [546]:
class GrandParent:
    def __init__(self):
        print("Hello GrandParent")


class Parent(GrandParent):
    def __init__(self):
        super().__init__()
        print("Hello Parent")


# class Child(GrandParent, Parent):
#     def __init__(self):
#         super().__init__()
#         print("Hello Child")

# c = Child()

这里直接会报错，因为Child不能同时继承GrandParent和Parent

In [547]:
# class A:
#     a = 42
#     b = list(a + i for i in range(10))
#     print(b)
#
# A()

In [548]:
import re
x = "abaax12323ddd"
t = re.match(r"[a-zA-Z]+(\d+?)[a-zA-Z]+", x).group(1)
print(t)
t = re.match(r"[a-zA-Z]+(\d+)[a-zA-Z]+", x).group(1)
print(t)
t = re.match(r"[a-zA-Z]+(\d*)[a-zA-Z]+", x).group(1)
print(t)
t = re.match(r"[a-zA-Z]+(\d+?)", x).group(1)
print(t)

12323
12323
12323
1


group(0)表示整个匹配结果，group(1)表示第一个括号内的匹配值
显而易见[a-zA-Z]+表示1个或者多个字母
第二和第三个没区别不讲
一和四的区别在于中间尾部的匹配，虽然都使用了非贪婪模式，但是第一中为了匹配后面的[a-zA-Z]+将不得不匹配全部内容
匹配为“12323“ ”ddd”
第四的区别在于由于不需要匹配最后部分，“12323ddd”将直接因为非贪婪变为“1”


In [549]:
class A:
    x = 0

    def func(self):
        print(self.x)


a = A()
a.x = 666
a.func()
# a.func(a)
A.func(a)
A.func(A)

666
666
0


"""
则下列哪些语句可以打印出0（ACD）

A、a.func()

B、a.func(a)

C、A.func(a)

D、A.func(A)
"""

```python
class A:
    x = 0
```
x = 0在这里叫做类变量，类变量是属于类本身的变量，而不是属于类的实例。这意味着，所有 A 类的实例都共享同一个 x 变量
a.x = 0 将实例a的x值进行了修改
a.func()，调用a实例x的值，由于已经定义了a.x = 0，输出0
a.func(a)，报错，因为不需要参数
A.func(a) 相当于a.func()
A.func(A)


In [550]:
class Foo:
    count = 0

    def __init__(self):
        self.count = 0

    def incr(self):
        self.count += 1

    @staticmethod
    def incr_static():
        Foo.count += 1

    @classmethod
    def incr_class(cls):
        print(str(cls))
        cls.count += 1


class Bar(Foo):
    pass

foo = Foo()
bar = Bar()
bar.incr()
bar.incr_static()
bar.incr_class()

<class '__main__.Bar'>


A、foo.count的值为1

B、bar.count的值为1

C、Bar.count的值为1

D、Foo.count的值为1

In [551]:
print(foo.count,bar.count,Bar.count,Foo.count)

0 1 2 1


Bar继承了Foo，但是Bar本身没有定义任何数据，因此所有的修改都会修改Foo
bar.incr(),会将实例化的bar的count值加1
bar.incr_static() 会将Foo的类变量count加1
bar.incr_class() 由于这里使用了cls，这里的cls将专门指代Bar，而不是Foo，因此Bar的类变量count会加1

有如下几个文件，目录结构及文件内容如下：
```
.
├── test1.py
└── tests
    ├── __init__.py
    └── test2.py
```
test1.py:
```python
from tests import test2

test2.test_2()
```
tests/__init__.py:
```python
import test2

def test2():
    print('hello python')
```
tests/test2.py:
```python
def test_2():
    print('hello world')
```
运行test1.py的结果是（C）

A、hello python

B、hello world

C、AttributeError: 'function' object has no attribute 'test_2

D、AttributeError: module 'tests' has no attribute 'test2

这道题涉及python导入优先级，python会优先去__init__.py 文件中查找名为 test2 的变量、函数或类
如果找到了，就会导入它。如果没有找到，解释器会继续查找名为 test2 的子模块或子包。

自定义的异常类型必须继承自Exception，按照已知异常->Exception的顺序捕获时，不需要关注对键盘中断、进程退出、生成器结束等异常的特殊处理。
禁止使用非异常类型作为raise对象，如NotImplemented、可迭代对象等。
下面是一个正例子：

In [552]:
class AuthenticationError(Exception): # 使用Exception而不是BaseException
    def __init__(self, message=None):
        super(AuthenticationError, self).__init__(message)
        self.http_code = 404


def handle_request(req):
    try:
        ...
    except AuthenticationError as e:
        return Response(e.http_code, e.message)
    except Exception as e:  # AuthenticationError等异常会正常抛出，无需额外关注
        return Response(400, e.message)

In [553]:
# 模块内全局变量
_global = object()

# 全局可见的全局变量
overall_global = object()

# 全局常量
CONSTANT = 60 * 60

下述哪些选项的说法符合《华为Python语言编程规范》中“变量作用域”章节的原则、要求和建议（ABC）
```python
__all__ = []
_g_type = 1


def get_type():
    global _g_type
    return _g_type


def set_type(type):
    _g_type = type
```
A、get_type函数中的global的使用违反规范

B、set_type函数中的type参数命名违反规范

C、set_type函数中的_g_type的命名违反规范

D、全局变量_g_type的命名违反规范，模块内使用的全局变量不能以下划线开头



考察：G.VAR.02 禁止使用global关键字声明不存在的外部变量
A 正确，
get_type 是直接返回了_g_type，在读取外部变量的时候，不需要使用global进行声明
B 正确
type是python自带函数，不建议使用
_g_type = type会覆盖外部变量_g_type的值
C 正确
如果必要使用全局变量，应该通过命名（以_开头）将全局变量声明为模块的内部变量，并且将其排除在__all__之外。
D 错误
见C


In [554]:
import datetime
d2 = datetime.datetime(5, 6, 7, 8, 9, 10)
print(d2.year, " Q", d2.month // 4 + 1, sep='')

5 Q2


datetime(year, month, day, hour=0, minute=0, second=0, microsecond=0, tzinfo=None, *, fold=0)
The year, month and day arguments are required

year=5,month=6 代入即可

In [555]:
language = u"中文"
print(len(language), hasattr(language, "encode"))

2 True


u"中文"是Python 2.x中的Unicode字符串，其中的u表示该字符串是Unicode编码。在Python 3.x中，所有字符串都是Unicode字符串，因此不需要使用u前缀。

下述哪个选项的代码符合《华为Python语言编程规范》中“运算符”章节的原则、要求和建议（C）

A、比较两个list内容是否相同：

if list1 is list2:
B、比较两个时间是否相等：

if time1 is time2:
C、比较两个tuple的内容是否相同：

if tuple1 == tuple2:
D、判断tuple内容是否为空：

if my_tuple is ():

G.OPR.03 禁止使用is或is not运算符在内置类型之间作比较，建议用==或!=

如下哪些程序执行能输出attr_name的值（ABD）
```

A、
class Parent:
    attr_name = 1

Foo = type('', (Parent,), {})
print(Foo.attr_name)
B、
class Parent:
    attr_name = 1

Foo = type('', (), {'attr_name': 1})
foo = Foo()
print(foo.attr_name)
C、

class Parent:
    attr_name = 1


Foo = type('Parent', (), {})
print(Foo.attr_name)

D、
class Parent:
    attr_name = 1

Foo = type('', (Parent,), {})
foo = Foo()
print(foo.attr_name)
```

Foo = type('', (Parent,), {})
这是一个Python中的动态类创建方法。type()函数可以接受三个参数，分别是type(name, bases, dict)。其中，name是类名，bases是继承的父类集合，dict是属性集合。这个语句的意思是创建一个名为Foo的类，它继承自Parent类，但没有任何属性¹。
A 正确
这里相当于创建了一个继承Parent的Foo类，同时也继承了Parent的类变量，输出1
B
这里的Foo没有继承任何类，但是定义了attr_name属性，能输出1
C
无继承，无属性，无结果
D
输出实例foo的attr_name属性


In [556]:
class Parent:
    attr_name = 1


Foo = type('', (), {'attr_name': 1})
foo = Foo()
print(foo.attr_name)

1


In [557]:
def decorate_function(function):
    def inner(*args, **kwargs):
        inner.co += 1
        return function(*args, **kwargs)

    inner.co = 0
    return inner


@decorate_function
def sample_function():
    pass


sample_function()
sample_function()
sample_function()
print(sample_function.co)

3


`sample_function` 有 `co` 这个属性是因为在定义装饰器函数 `decorate_function` 时，新函数 `inner` 被定义为一个闭包，它可以访问外部函数的变量。在这种情况下，`inner` 函数可以访问 `decorate_function` 函数中定义的变量 `co`。每次调用 `sample_function` 函数时，实际上是调用了 `inner` 函数，因此每次调用都会增加 `co` 变量的值。由于 `co` 变量是作为 `inner` 函数的属性存储的，因此它可以跨多个调用保持不变。

下面的几个删除目录及目录下所有文件的操作中，符合《华为Python语言编程规范》中“代码工程”章节的原则、要求和建议的是（D）

A、

os.system(f"rm -r {dir_name}")
B、

proc = subprocess.Popen(['rm', '-r', dir_name], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
outs, errs = proc.communicate()
C、

subprocess.run(['rm', '-r', dir_name])
D、

shutil.rmtree(dir_name)

A 错误
P.11 建议尽量使用标准库中的功能来替代直接调用操作系统的系统命令
os.remove shutil.rmtree可以实现删除功能


如下代码段，输出字符串为“海贝色”的是(D )

DictColor = {"seashell": "海贝色", "gold": "金色", "pink": "粉红色", "brown": "棕色", "purple": "紫色", "tomato": "西红柿色"}
A、

print(DictColor.keys())
B、

print(DictColor["海贝色"])
C、

print(DictColor.values())
D、

print(DictColor["seashell"])

基础题 跳过

In [558]:
def decorate(func):
    def wrapper():
        print('start')
        func()
        print('end')

    return wrapper


@decorate
def show():
    return 'show'


print(show())

start
end
None


decorate(show),很显然 print('start') func() print('end')，输出 start end，wrapper未定义返回值

已有变量定义 name = "tom"，则下列表达式中值为字符串"tom"的是（C）

A、

"{name!r}"


B、

"{name%r}"
C、

f"{name!r}"
D、

f"{name%r}"

选项C的表达式是f"{name!r}"，它的值为字符串"tom"。在这个表达式中，f表示这是一个格式化字符串，、
{name!r}表示将变量name的值转换为字符串并使用repr()函数来表示它。repr()函数返回对象的 string 格式
除了r还可以使用s和a
!a表示使用ascii()函数来表示变量的值，!s表示使用str()函数来表示变量的值


In [559]:
name = "tom"
"{name!r}","{name%r}",f"{name!r}" # f"{name%r}"报错

('{name!r}', '{name%r}', "'tom'")

下面all函数对字典的操作，输出为True的有（）下面all函数对字典的操作，输出为True的有（BD）

A、

print(all({0: 'True', 1: 'True'}))
B、

print(all({1: 'False', 2: 'True'}))
C、

print(all({1: 'True', False: 2}))
D、

print(all({'0': 'True'}))

all() 函数用于判断给定的可迭代参数 iterable 中的所有元素是否都为 TRUE，如果是返回 True，否则返回 False。

AD:
字典对象将检查所有键值，第一个键值为0，返回false
BC:
键值全部不为0，返回True

下列构造文件路径的方式，符合《华为Python语言编程规范》中“文件”章节的要求和建议的有（BC）

A、

import os


file_name = 'tmp.txt'
file_path = os.getcwd() + '\\' + file_name
B、

import os


file_name = 'tmp.txt'
file_path = os.getcwd() + os.sep + file_name
C、

from pathlib import Path


file_name = 'tmp.txt'
file_path = Path(Path.cwd(), file_name)
D、

from pathlib import Path


file_name = 'tmp.txt'
file_path = Path(Path.cwd()) / file_name

A:
G.FIO.05 构造文件路径时应屏蔽操作系统的差异，建议使用库函数辅助构造
使用'\'进行路径拼接会导致其路径在win系统无法使用
BC:【正例】包含，使用OS和pathlib函数构造路径
D:
同A


以下哪些变量定义语句执行时会产生异常？（CD ）

A、

a = 100, 200, 300
B、

a = b = c = 10
C、

a, b = 100, 300, 500
D、

a = 10 20

A选项
这种写法会自动转换为一个元组，不会报错

以下哪些选项输出结果为True ( AB)

A、

print(2 == 0 or True)
B、

print(2 != 0 or True)
C、

print(str(2) and str(0) or True)
D、

print(str(2) or str(0) or True)

A
Fales or True -> True
B
True or True -> True
C
'0' or True -> '0'
D
'2' or True -> '2'

'' or True -> True
'' or False -> ''

下列哪些选项符合《华为Python语言编程规范》中“类与面向对象”章节的原则、要求和建议（ACD）
```
A、

class Summation:
    def __init__(self, base):
        self._base = base

    def add(self, a: int):
        self._base += a
B、

class Addition:
    def add(self, a: int, b: int) -> int:
        return a + b
C、

class Addition:
    @staticmethod
    def add(a: int, b: int) -> int:
        return a + b
D、

from abc import ABC, abstractmethod


class Base(ABC):
    def __init__(self, base):
        self._base = base

    @abstractmethod
    def cal(self, a, b):
        pass


class Addition(Base):
    def cal(self, a, b):
        return a + b
```

B:
G.CLS.07 类的方法不需要访问实例时，建议定义为staticmethod或classmethod
D:
G.CLS.04 建议在实现抽象类时使用abc模块



抽象类是面向对象程序设计中的一个重要概念。它是一种特殊的类，只能被继承，不能被实例化。抽象类通常包含一个或多个抽象方法，这些方法只有声明，没有具体的实现。子类必须实现这些抽象方法，否则它们也将成为抽象类。

在 Python 中，可以使用 `abc` 模块来定义抽象类。例如，下面是一个简单的抽象类的例子：

```python
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

    @abstractmethod
    def perimeter(self):
        pass
```

在这个例子中，我们定义了一个名为 `Shape` 的抽象类，它包含两个抽象方法：`area` 和 `perimeter`。这些方法没有具体的实现，因此子类必须实现它们。

总之，抽象类是一种特殊的类，它提供了一种定义接口的方式。子类必须实现抽象类中的所有抽象方法，否则它们也将成为抽象类。这样可以确保子类遵循特定的接口规范。

下面哪些表达式结果的数据类型为int（ABD）

A、

0xacd
B、

12_00
C、

float(3)
D、

0b1011



- 表达式 A：`0xacd` 是一个十六进制整数，它的十进制值为 `2765`。
- 表达式 B：`12_00` 是一个整数，其中的下划线只是为了提高可读性，它的值为 `1200`。
- 表达式 C：`float(3)` 将整数 `3` 转换为浮点数，结果为 `3.0`。
- 表达式 D：`0b1011` 是一个二进制整数，它的十进制值为 `11`。


下面哪些表达式操作可以将x = ['a', 'b'] 转变为['a', 'b', 'c', 'd']（ABD）

A、

x.extend(['c', 'd'])
B、

x += ['c', 'd']
C、

x.append(['c', 'd'])
D、

x += "cd"

D:
[+=可以视为串联赋值](https://python-reference.readthedocs.io/en/latest/docs/operators/concatenation_assignment.html)

关于下面的代码的输出，正确的是（B）

language = b"Chinese"
print(language)
print(type(language))
print(str(language))
print(type(str(language)))

In [560]:
language = b"Chinese"
print(language)
print(type(language))
print(str(language))
print(type(str(language)))

b'Chinese'
<class 'bytes'>
b'Chinese'
<class 'str'>


1. b"Chinese"为字节串而不是字符串，字节串的print输出都会以b开头，编码形式为ASCII
2. str函数直接对字节串进行转换的时候，默认为ASCII编码，不会有任何变化
3. 如果想要输出Chinese，那么str(language, encoding='utf-8')，指定编码方式即可


下述哪个选项的代码符合《华为Python语言编程规范》中“函数与方法”章节的要求和建议（D）
```
A、

def division(x, y):
    if y != 0:
        print(x / y)

result = division(1, 2)
B、

def division(x, y):
    if abs(y) > 1e-5:
        print(x / y)

result = division(1, 2)
C、

def division(x, y):
    try:
        print(x / y)
    except ZeroDivisionError:
        print('除数不能为0')

result = division(1, 2)
D、

def division(x, y):
    if y != 0:
        print(x / y)

division(1, 2)
```

G.FNM.04 不要将没有返回值的函数调用结果赋值给变量
division返回为空值，直接执行即可，不应赋值给新变量

此外
G.EXP.01 尽可能使用隐式布尔值
if y != 0使用if y代替更好
从这个角度来看，是没有正确选项的
推荐写法如下
```python
def division(x, y):
    try:
        print(x / y)
    except ZeroDivisionError:
        return False
    return True

result = division(1, 2)
```




下列哪个选项符合《华为Python语言编程规范》中“类与面向对象”章节的原则、要求和建议（C）


```
A、

import sys


class Base:
    def __init__(self):
        self.running = False

    def start(self):
        self.running = True


class Addition(Base):
    def start(self, a, b):
        super().start()
        result = a + b
        sys.stdout.write(str(result))
B、

import sys


class Base:
    def __init__(self):
        self.running = False

    def start(self, *args):
        self.running = True


class Addition(Base):
    def __init__(self):
        self.running = False

    def start(self, *args):
        super().start(*args)
        result = sum(*args)
        sys.stdout.write(str(result))
C、

import sys


class Base:
    def __init__(self):
        self.running = False

    def start(self, *args):
        self.running = True


class Addition(Base):
    def start(self, *args):
        super().start(*args)
        result = sum(*args)
        sys.stdout.write(str(result))
D、

import sys


class Base:
    def __init__(self):
        self.running = False

    def start(self, *args):
        self.running = True


class Addition(Base):
    def __init__(self):
        Base.__init__(self)
        self.running = True
```

A错误
于其他函数相比，直接相加局限性太大（个人理解）
B
G.CLS.01 反例
父类已经定义running的情况下，重复定义会导致父类初始化过程遗漏
class Base:
    def __init__(self):
        super.__init__()
        self.running = False
D
G.CLS.05 子类调用父类初始化方法的方式，建议使用super类辅助
避免菱形继承

In [561]:
def func():
    yield (x for x in range(3))


for x in func():
    print(tuple(x))

(0, 1, 2)


```
class Shape:
    def __init__(self, name):
        self.name = name


class Rectangle(Shape):
    def __init__(self, length, width):
        super().__init__('Rectangle')
        self._length = length
        self._width = width

    def area(self):
        return self._length * self._width


class Square(Rectangle):
    def __init__(self, length):
    ______________ # 请从选项中选择合适的代码

square = Square(5)
print(square.name)
```
下列哪些选项中的代码单独填入到横线处时，可以使得以上代码执行后的输出结果为Square？（ ACD）

A、

super().__init__(length, length)
Shape.__init__(self, 'Square')
 B、

self._name = 'Square'
super().__init__(length, length)
 C、

super(Square, self).__init__(length, length)
super(Rectangle, self).__init__('Square')
 D、

super().__init__(length, length)
super(Rectangle, self).__init__('Square')

继承顺序Square->Rectangle->Shape
注意到Rectangle 的init函数会将name命名为Rectangle

A:正确
Shape.__init__(self, 'Square')直接调用父类并指定名称
B:错误
self._name 不是name变量，不改变结果，依旧输出Rectangle
C:正确
super(Rectangle, self).__init__('Square')
指的是调用Rectangle父类的方法，也就是shape的init方法
D:正确
于C区别不大


以下哪些选项的代码执行之后，s是一个包含'a', 'b', 'c'三个字符元素的集合（set）（ABE ）
A、

s = set('abc')
 B、

s = {'a', 'b', 'c'}
 C、

s = {('a', 'b', 'c')}
 D、

s = set('a', 'b', 'c')
 E、

s = set(['a', 'b', 'c'])

A:
set 函数接受一个可迭代对象作为参数，并从中提取元素以创建集合。字符串是一个可迭代对象,所以会正确输出
B:
集合的定义
C:
这是一个单元素集合
D:
报错，对象数量超过1个
E:
见A

以下含有多行的代码，可以正常执行的有（BCD ）
```
A.

 s = 'a' + 'b'
     + 'c' + 'd'
     + 'e' + 'f'
B.

 s = 'a' + 'b' \
     + 'c' + 'd' \
     + 'e' + 'f'
C.

 s = ('a' + 'b'
     + 'c' + 'd'
     + 'e' + 'f')
D.

 s = 'a' + 'b' \
     + 'c' + 'd' \
     + 'e' + 'f' \
```

反斜杠（\）用于在 Python 中表示续行,末尾有不会报错

在linux环境中，为了能够import在/my_path路径下的包(package)，以下做法不符合《华为Python语言编程规范》中“标准库”章节的要求和建议的有（BCD）

A.

运行代码前，在bash中执行export PYTHONPATH=$PYTHONPATH:/my_path

B.

在代码中添加

 sys.path.insert(0, './')
 sys.path.insert(0, './my_path')
C.

在代码中添加sys.path = ['/my_path']

D.

在代码中添加sys.path.insert(0, '/my_path')

G.PSL.03 修改sys.path时注意避免模块重名问题

1. 在修改sys.path时注意避免模块重名导致的导入失败，建议使用sys.path.append代替sys.path.insert(0, my_directory)。
2. 除此之外，也可以在运行python进程前，修改系统PYTHONPATH环境变量达到相同的目的。
3. 建议以项目或子项目的根目录作为待添加入sys.path的路径

有如下的文件结构：
```
 .
 ├── test1.py
 └── tests
     ├── __init__.py
     ├── test2.py
     └── test3.py
在test2.py中，以下几个包导入操作可以正确执行的有（AB）

A.

from . import test3
B.

from . import *
C.

from ..tests import test3
D.

import test3
```

ABC均为从相对目录导入，但是C是从上级目录导入，题目中当前目录为tests文件夹内，所以AB成功，C失败
D属于直接从系统目录导入

In [562]:
class Foo:
    def __init__(self, name):
        self._name = name

    def __str__(self):
        return self._name

    def __eq__(self, other):
        print('foo', self, other)
        return self is other


class Bar(Foo):
    def __eq__(self, other):
        print('bar', self, other)
        return str(self) == str(other)


foo = Foo('Hello')
bar = Bar('Hello')
print(foo == bar)
print(bar == foo)

bar Hello Hello
True
bar Hello Hello
True


- 上面这段代码，Bar继承了Foo，并重新定义了__eq__，因此，Foo的__eq__会被Bar的__eq__重写（Method Overriding），无论调用父类还是子类，执行的都是子类的方法。
- 需要注意的是，这个是__eq__函数的特殊情况，因为__eq__同时涉及到了两个类，常规情况下，只声明Foo，其同名函数并不会被重写

In [563]:
def func(x):
    x = yield x
    x = yield x


m = func(3)
for x in m:
    print(x, end=' ')
    m.send(0)

3 

如下表达式的值为False的有(ACD)

 A、{3, 5, 4} < {4, 3, 5}

 B、{5, 3} < {4, 3, 5}

 C、{3, 5, 4} > {4, 3, 5}

 D、{3, 5, 6} > {4, 3, 5}
 <用于判断集合是否为其子集

在 Python 中，`<` 运算符用于比较两个对象的值。如果第一个对象小于第二个对象，则返回 `True`。否则，返回 `False`。但是，当 `<` 运算符用于比较不同类型的对象时，它的行为可能会有所不同。

- 对于数字类型（如整数和浮点数），`<` 运算符按照数学规则进行比较。例如，表达式 `3 < 4.0` 返回 `True`，因为整数 3 小于浮点数 4.0。

```python
print(3 < 4.0) # True
```

- 对于字符串类型，`<` 运算符按照字典顺序进行比较。例如，表达式 `'cat' < 'dog'` 返回 `True`，因为字符串 `'cat'` 在字典中排在字符串 `'dog'` 之前。

```python
print('cat' < 'dog') # True
```

- 对于列表和元组类型，`<` 运算符按照元素顺序进行比较。例如，表达式 `[1, 2] < [1, 3]` 返回 `True`，因为列表 `[1, 2]` 的第二个元素小于列表 `[1, 3]` 的第二个元素。

```python
print([1, 2] < [1, 3]) # True
```

- 对于集合类型，`<` 运算符用于判断一个集合是否是另一个集合的子集。例如，表达式 `{1, 2} < {1, 2, 3}` 返回 `True`，因为集合 `{1, 2}` 是集合 `{1, 2, 3}` 的子集。

```python
print({1, 2} < {1, 2, 3}) # True
```

然而，当 `<` 运算符用于比较不同类型的对象时，它的行为可能会引发 `TypeError`。例如，表达式 `3 < 'cat'` 将引发错误，因为整数和字符串不能直接进行比较。

```python
try:
    print(3 < 'cat')
except TypeError as e:
    print(e) # '<' not supported between instances of 'int' and 'str'
```

In [564]:
# 以下代码的输出结果是（ B）

class Foo:
    @classmethod
    def create_one(cls):
        return cls()

    @staticmethod
    def create_two():
        return Foo()


class Bar(Foo):
    pass


bar_one = Bar.create_one()
bar_two = Bar.create_two()
print(isinstance(bar_one, Bar))
print(isinstance(bar_two, Bar))

True
False



`bar_one` 是通过 `Bar.create_one()` 创建的，它是一个 `Bar` 类的实例。`create_one` 是一个类方法，它接收类本身作为第一个参数，因此 `cls()` 等价于 `Bar()`。

`bar_two` 是通过 `Bar.create_two()` 创建的，它是一个 `Foo` 类的实例。`create_two` 是一个静态方法，它不接收类本身作为参数，因此 `Foo()` 总是创建一个 `Foo` 类的实例，而不管它是从哪个类调用的。

因此，`isinstance(bar_one, Bar)` 返回 `True`，而 `isinstance(bar_two, Bar)` 返回 `False`。

下述哪个选项的代码符合《华为Python语言编程规范》中“函数与方法”章节的要求和建议（B）
```
A、

def fun(arg_0, arg_1='', arg_2=[]):
    arg_2.append(arg_0)
    arg_2.append(arg_1)
    print(arg_2)
B、

def fun(arg_0, arg_1='', arg_2=None):
    arg_2 = [] if arg_2 is None else arg_2
    arg_2.append(arg_0)
    arg_2.append(arg_1)
    print(arg_2)
C、

def fun(arg_0, arg_1='', arg_2={}):
    arg_2.update({"arg_0": arg_0})
    arg_2.update({"arg_1": arg_1})
    print(arg_2)
D、

def fun(arg_0, arg_1='', arg_2=set()):
    arg_2.add(arg_0)
    arg_2.add(arg_1)
    print(arg_2)
```

G.FNM.01 禁止使用可变对象作为参数默认值
字典，集合，列表均为可变对象

In [565]:
def func(a, *b):
    for item in b:
        a += item
    return a


m = 0
print(func(m, 1, 1))

2


现有numbers = [1, 2, 3]，则以下代码哪些的运行结果是[2]（BD）

A、

print(list(filter(lambda x: x > 1, numbers)))
B、

print(list(filter(lambda x: x % 2 == 0, numbers)))
C、

print(list(filter(lambda x: 2, numbers)))
D、

print(list(filter(lambda x: (x + 1) % 3 == 0, numbers)))


filter函数会将所有lambda函数返回为真的留下来
A [2,3]
B 2
C [1,2,3]
D [2]

以下哪个选项符合《华为Python语言编程规范》中“异常处理”章节的要求和建议（C）

```

A、

try:
    do_something()
except SomeException:
    raise
except Exception as e:
    do_something(e)
B、

def control_statement_in_finally(param):
    ret = -1
    try:
        ret = 1 / param
    finally:
        return ret
C、

def control_statement_in_finally(param):
    ret = -1
    try:
        ret = 1 / param
    except ZeroDivisionError:
        do_something()
    finally:
        do_finally()
    return ret
D、

try:
    fp = open('test_read.txt', encoding='utf8')
    print(fp.read())
except FileNotFoundError:
    pass
except UnicodeDecodeError:
    pass
finally:
    fp.close()
```

A
G.ERR.06 raise语句必须包含异常实例
raise 如果不返回异常，则会默认抛出RuntimeError，丢失了原始错误调用栈信息。
B
G.ERR.14不要使用return、break、continue或抛出异常使finally块非正常结束
try后未对错误进行处理，导致即使出现错误也会继续执行finally语句，导致无法捕获错误
D
G.ERR.07 避免抑制或忽略异常
编码时应尽量避免捕获一个异常但什么都不做的情况

已知：x = [ 4, 1, 0, 3, 5]， 下面哪些表达式操作可以将 x转变为[ 4, 1, 3, 5]（BCD）
```
A、

x[2] = []
B、

del x[2]
C、

x.remove(0)
D、

x[2:3] = []
```

In [566]:
class A:
    def __new__(self):
        print("A's __new__() invoked")
        return object.__new__(self)

    def __init__(self):
        print("A's __init__() invoked")


class B(A):
    def __new__(self):
        print("B's __new__() invoked")

    def __init__(self):
        print("B's __init__() invoked")


obj1 = B()
obj2 = A()

B's __new__() invoked
A's __new__() invoked
A's __init__() invoked


请认真阅读这段代码
这里的B类继承了A，

- object 类是 Python 中所有类的基类，它提供了一些基本方法，包括 __new__ 方法
- __new__ 方法是一个静态方法，它负责创建并返回一个新的对象
- 因此A类的__new__返回自己为对象的类
- 但是对于B类，由于B类并未返回任何值，所以B类将无法执行__init__函数（相当于重写了object的new）
- __new__先于__init__执行，这里obj1将只输出B's __new__() invoked
- obj2将正常执行